mm2src/mm2_bitcoin/spv_validation/src/helpers_validation.rs

```diff
@@ -378,13 +394,139 @@ pub async fn validate_headers(
+        previous_header = SPVBlockHeader::from_block_header_and_height(header, previous_height + 1);
         previous_hash = previous_header.hash;
         previous_height += 1;
     }
     Ok(())
 }
 
+#[derive(Debug, Display)]
+pub(crate) enum ChainReorgError {
+    #[display(fmt = "Can't get from the storage for {} - reason: {}", coin, reason)]
+    BlockNotFound { coin: String, reason: String },
+    #[display(fmt = "Invalid chain detected for {} new block headers at height : {}", coin, height)]
+    InvalidChain { coin: String, height: u64 },
+    #[display(
+        fmt = "Unable to delete block headers from_height{from_height} to_height {to_height} from storage for \
+    {coin}\
+     - reason: \
+    {reason}"
+    )]
+    UnableToDeleteHeaders {
+        from_height: u64,
+        to_height: u64,
+        coin: String,
+        reason: String,
+    },
+}
+
+impl From<ChainReorgError> for SPVError {
+    fn from(value: ChainReorgError) -> Self { Self::ChainReorgError(value.to_string()) }
+}
+
+// Check for a chain reorg given a block header and a list of headers
+// downloaded from an RPC node and remove bad headers from storage by block difficulty.
+pub(crate) async fn detect_and_resolve_conflicting_headers_by_difficulty(
+    coin: &str,
+    previous_height: u64,
+    previous_header: &SPVBlockHeader,
+    last_block_height: u64,
+    new_header: &BlockHeader,
+    storage: &dyn BlockHeaderStorageOps,
+) -> Result<(), ChainReorgError> {
+    if previous_header.previous_header_hash == new_header.previous_header_hash
+        && (previous_header.hash != new_header.hash() && previous_header.bits != new_header.bits)
+    {
+        if u32::from(previous_header.bits.clone()) > u32::from(new_header.bits.clone()) {
+            println!(
+                "Chain reorg detected at height: {last_block_height} -> hash: {} -> coin: {coin}",
+                new_header.hash()
+            );
+            // Todo: retry header sync using another eletrum.
+            return Err(ChainReorgError::InvalidChain {
+                coin: coin.to_string(),
+                height: previous_height + 1,
+            });
+        } else {
+            // Remove bad conflicting header and successors from storage.
+            storage
+                .remove_headers_from_storage(previous_height + 1, last_block_height)
+                .await
+                .map_err(|err| ChainReorgError::UnableToDeleteHeaders {
+                    from_height: previous_height + 1,
+                    to_height: last_block_height,
+                    coin: coin.to_string(),
+                    reason: err.to_string(),
+                })?;
+        };
+    }
+
+    Ok(())
+}
+
+// Check for a chain reorg given a block header and a list of headers
+// downloaded from an RPC node and remove bad headers from storage by block hash.
+pub(crate) async fn detect_and_resolve_conflicting_headers_by_hash(
```

Comment:
I think we should only have one function to detect chain reorg, there are a lot of duplicate code in this and `detect_and_resolve_conflicting_headers_by_difficulty`

Reply:
done